etdumpfilter impl

Gasoonjia opened the pull request:
Differential Revision: D72025517

Gasoonjia commented:
@pytorchbot label "topic: not user facing"

swolchok commented on devtools/etdump/etdump_filter.cpp:
```diff
@@ -0,0 +1,101 @@
+using ::executorch::runtime::Error;
+using ::executorch::runtime::kUnsetDelegateDebugIntId;
+
+namespace executorch {
+namespace etdump {
```
we have C++17 -- `namespace executorch::etdump {`

swolchok commented on devtools/etdump/etdump_filter.h:
```diff
@@ -0,0 +1,107 @@
+
+ private:
+  char regex_patterns_[MAX_REGEX_PATTERNS][MAX_PATTERN_LENGTH]{};
+  size_t regex_count_;
+  size_t range_start_;
+  size_t range_end_;
```
nit: you don't need the explicit default constructor if you assign 0 to these here; you can then change the constructor implementation in the .cpp file to `ETDumpFilter::ETDumpFilter() = default;`

swolchok commented on devtools/etdump/etdump_filter.cpp:
```diff
@@ -0,0 +1,101 @@
+  if (len >= MAX_PATTERN_LENGTH) {
+    return Error::InvalidArgument; // Pattern too long
+  }
+  strcpy(regex_patterns_[regex_count_], pattern);
```
nit: you already have the length, so use it -- `std::memcpy(regex_patterns_[regex_count_], pattern, len)`

swolchok commented on devtools/etdump/etdump_filter.cpp:
```diff
@@ -0,0 +1,101 @@
+    return Error::InvalidArgument; // Pattern too long
+  }
+  strcpy(regex_patterns_[regex_count_], pattern);
+  regex_patterns_[regex_count_][len] = '\0';
```
FYI strcpy already does this. however, if you take my memcpy advice above you'll need to leave it

swolchok commented on devtools/etdump/etdump_filter.cpp:
```diff
@@ -0,0 +1,101 @@
+  return true;
+}
+
+Result<bool> ETDumpFilter::set_debug_handle_range(size_t start, size_t end) {
```
what does a successful result of `false` mean? (does `Result<void>` work?)

Gasoonjia replied:
false will not be use. 
We can not use `Result<void>` cuz Result class need to use reference for the T type and void does not have Reference.

swolchok commented on devtools/etdump/etdump_filter.cpp:
```diff
@@ -0,0 +1,101 @@
+
+Result<bool> ETDumpFilter::filter_name_(const char* name) {
+  if (name == nullptr) {
+    return Error::InvalidArgument; // Name is null
```
comment repeats the code, delete

swolchok commented on devtools/etdump/etdump_filter.h:
```diff
@@ -0,0 +1,107 @@
+  size_t get_n_regex() const;
+
+ private:
+  char regex_patterns_[MAX_REGEX_PATTERNS][MAX_PATTERN_LENGTH]{};
```
why the hardcoded limit on length instead of a `std::vector<std::string>`? I don't see any particular guarantees from RE2 itself that it doesn't malloc, so isn't studiously avoiding malloc in our wrapper a lost cause?

Gasoonjia replied:
I don't want to use any memory allocation operation in etdump_filter; our user usually have their own memory allocation way and does not support anything like vector or malloc.

swolchok replied:
> does not support anything like vector or malloc

but that's my point -- RE2 mallocs! (for example https://github.com/google/re2/blob/main/re2/compile.cc#L1254)

Gasoonjia replied:
ahhh fair point didn't pay attention to RE internal impl

swolchok commented on devtools/etdump/etdump_filter.h:
```diff
@@ -0,0 +1,107 @@
+   *         - An error code if number of pattern has reached to cap, or any
+   * error occurs during regex compilation.
+   */
+  Result<bool> add_regex(const char* pattern);
```
std::string_view is in C++17. why not use it instead? (if the user ever passes a literal, we won't need to check its length at runtime)

Gasoonjia replied:
thx didn't know we have that; migrating to that.

swolchok reviewed (commented):
just some C++ comments; leaving for reviewers with more context on the actual goals here

tarun292 commented on devtools/etdump/etdump_filter.cpp:
```diff
@@ -0,0 +1,100 @@
+ETDumpFilter::ETDumpFilter()
+    : regex_count_(0), range_start_(0), range_end_(0) {}
+
+Result<bool> ETDumpFilter::add_regex(const char* pattern) {
```
I think let's call this add_filter, we don't necessarily want to tie this to regex i think. On smaller platforms where they pass their own filter implementation they can just do simple string matching.

tarun292 commented on devtools/etdump/etdump_filter.cpp:
```diff
@@ -0,0 +1,100 @@
+/*
```
Overall high level comment, i think it's reasonable that we can assume the etdump filter we provide based on regex will need malloc to be supported. If users want to use this malloc support will be required on their platform. Based on this i think it would be good to remove the limitations we have in this class such as max regex, max char len etc.

tarun292 replied:
If users do need a filter for their platform that isn't based on regex they can write their own one. (which will be only very few embedded users).

Gasoonjia replied:
> which will be only very few embedded users
ok  in this case i agree we can make our etdump_filter support regex + malloc, just as what [swolchok](https://github.com/swolchok) mentioned RE use malloc internally. Let me make an update

Gasoonjia committed:
etdumpfilter impl (#9752)

Summary:
Pull Request resolved: https://github.com/pytorch/executorch/pull/9752

Introduce ETDumpFilter for filtering logged delegation intermediate data.
Details can be found in https://github.com/pytorch/executorch/discussions/9260

Reviewed By: tarun292

Differential Revision: D72025517
